Feature/homework 6

requiemail opened the pull request:
Шестое домашнее задание от 01.04.2020.

Задание:

Переписать приложение для хранения книг на ORM

1. Использовать JPA, Hibernate только в качестве JPA-провайдера.
2. Добавить комментарии к книгам, и высокоуровневые сервисы, оставляющие комментарии к книгам.
3. Покрыть репозитории тестами

Commit:
Удалены DAO и их тесты, а также RowMappers

Commit:
Добавлена модель Comment, все модели размечены аннотациями @Entity

Commit:
Добавлены репозитории и их имплементации

Commit:
Внесены необходимые изменения в сервисы

Commit:
Добавлены необходимые изменения в скрипты liquibase

Commit:
Смена зависимости jdbc на jpa

Commit:
Отключена автогенерация

Commit:
Добавлено исключение

Commit:
Изменил выбрасываемое исключение на бизнес-исключение

Commit:
Убран лишний импорт

Commit:
Тесты сервисов

Commit:
Убраны лишние импорты

Commit:
Добавлены данные для заполнения таблицы comments

Commit:
Созданы тесты для репозиториев

Commit:
Мелкие правки

Commit:
Мелкие правки

Commit:
Исправлена некорректная DI

Commit:
Исправлены некорректные DI

kataus commented on bookshelf/src/main/java/ru/otus/homework/model/Book.java:
```diff
@@ -1,21 +1,72 @@
+@NamedEntityGraph(
+        name = "books-entity-graph",
+        attributeNodes = {
+                @NamedAttributeNode("comments"),
```
Комментарии нам нужны далеко не сразу. По факту они нужны только когда мы просматриваем карточку конкретной книги.

kataus commented on bookshelf/src/main/java/ru/otus/homework/model/Book.java:
```diff
@@ -27,4 +78,13 @@ public String toString() {
                 ", genres: " + genreList.stream().map(Genre::toString).collect(Collectors.joining(", "));
     }
 
+    public String toStringWithComments() {
```
Это особенность представления. Стоит там и делать, к самой сущности такой вывод на печать (как и любой вывод на печать) не имеет отношения

requiemail replied:
Согласен, вынесу.

kataus commented on bookshelf/src/main/java/ru/otus/homework/model/Comment.java:
```diff
@@ -0,0 +1,42 @@
+    private String commentText;
+    @Column(name = "comment_author")
+    private String commentAuthor;
+    @Column(name = "book_id")
```
Ой, зачем? Если связь есть - то нужное поле сделает JPA, а если нужно больше - то стоит сразу делать объект для связи

requiemail replied:
Не совсем понял что имеется в виду. Join-таблица в этом случае мне показалась избыточной, а хранить в комментариях внешний ключ, ссылающийся на книгу, мне показалось хорошей идеей. При этом таблицы у меня генерируются вручную, поэтому здесь я только название поля в таблице указываю. Если автогенерация ddl отключена, то как JPA сделает нужное поле?

kataus commented on bookshelf/src/main/java/ru/otus/homework/model/Genre.java:
```diff
@@ -1,18 +1,36 @@
     private String name;
 
     @Override
     public String toString() {
```
Мы используем везде ломбок, а тут сделали вручную. Стоит придерживаться единообразия. А если это для вывода на экран, то стоит это делать там где делается вывод на экран. Потому что выводов может быть много и каждому нужно своё

requiemail replied:
Да, этот аргумент мне в голову не приходил. Исправлю.

kataus commented on bookshelf/src/main/java/ru/otus/homework/repository/impl/AuthorRepositoryImpl.java:
```diff
@@ -0,0 +1,42 @@
+import java.util.List;
+import java.util.Optional;
+
+@Transactional
```
Ой, не надо ставить на весь класс, я же об этом говорил. Знаю что потом в примерах было на весь класс, но мы учли негативное впечатление от этого. Транзакции настоятельно рекомендуется делать точечно и там где это действиетльно нужно.

requiemail replied:
Да, я создал классы перед просмотром занятия о транзакциях, Александр Оруджев предлагал там проставлять аннотацию над классом сразу. Когда смотрел ваше занятие отметил себе, что есть другое мнение по этому вопросу, но не отметил необходимость рефакторинга в ДЗ. Сейчас займемся.

kataus commented on bookshelf/src/main/java/ru/otus/homework/repository/impl/AuthorRepositoryImpl.java:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    @Override
+    public Optional<Author> findById(long id) {
```
например тут транзакция с readolny = false мягко говоря избыточна. К тому же один запрос, find сам умеет открывать транзакцию

requiemail replied:
+

kataus commented on bookshelf/src/main/java/ru/otus/homework/repository/impl/BookRepositoryImpl.java:
```diff
@@ -0,0 +1,60 @@
+
+    @Override
+    public Optional<Book> findByIdWithComments(long id) {
+        EntityGraph<?> entityGraph = em.getEntityGraph("books-entity-graph");
```
Тут я бы назвал преждевременной оптимизацией. Круто что так сделали, но не настолько часто это надо чтобы заморачиваться, к тому же делать отдельный граф. Это пусть, в другом месте за связи спрошу

requiemail replied:
Проще было бы сделать через jpql запрос, включающий в себя fetch comment? Когда решал каким образом ленивую коллекцию подгружать в основном ориентировался на то, что мне будет интереснее реализовать, поэтому выбор остановил на графе.

kataus commented on bookshelf/src/main/java/ru/otus/homework/repository/impl/CommentRepositoryImpl.java:
```diff
@@ -0,0 +1,49 @@
+
+    @Override
+    public List<Comment> findAllByBookId(long id) {
+        TypedQuery<Comment> query = em.createQuery("from Comment where book_id = :id", Comment.class);
```
Вот так делать не стоит. У нас же есть book.getComments(). Потому данную задачу мы можем решить на уровне сервиса и даже не обращаясь к dao кроме как за книгой. давайте это точно подправим

requiemail replied:
Хорошо, убираем.

Commit:
Сохранение комментария реализовано через BookService, как часть книги

Commit:
Из Comment Entity убрано служебное поле book_id, убран за ненадобностью метод save() из репо, исправлены тесты

Commit:
Убран метод findAllByBookId() из репо и сервиса

Commit:
Получаем комментарии через BookService как часть книги

Commit:
Поправлен тест findById() для комментария, сохранение теперь реализовано каскадом через сохранение книги;

Commit:
toString() во всех entities реализованы через lombok, в ShellSessionRunner реализовано необходимое представление Book с подгруженной ленивой коллекцией.

Commit:
Решение возникающего org.hibernate.PersistentObjectException: detached entity passed to persist.
Возникает при попытке сохранить новую книгу с добавленными существующими в базе авторами или жанрами.

Коллекции авторов и жанров больше не обрабатываются каскадно, сохранение авторов и жанров происходит в BookService, выставлена @Transactional над методом save() в BookService.

Commit:
Убраны @Transactional над классами, добавлены необходимые @Transactional над методами.

Commit:
Добавлены @Transactional над save() методами

Commit:
Fix падающего теста(NPE из-за отсутствующих зависимостей AuthorService и GenreService)

Commit:
Ленивая коллекция подгружается через LEFT JOIN FETCH вместо EntityGraph

Commit:
Дополнены тесты, убраны лишние JOIN FETCH из BookRepositoryImpl

Commit:
В Comment добавлено поле Book, организована двусторонняя связь через mappedBy

kataus commented on bookshelf/src/main/java/ru/otus/homework/repository/impl/BookRepositoryImpl.java:
```diff
@@ -48,7 +48,7 @@ public void delete(Book book) {
     @Override
     public Optional<Book> findByIdWithComments(long id) {
-        TypedQuery<Book> query = em.createQuery("select b from Book b left join fetch b.comments left join fetch b.genreList left join fetch b.authorList where b.id = :id", Book.class);
+        TypedQuery<Book> query = em.createQuery("select b from Book b left join fetch b.comments where b.id = :id", Book.class);
```
Немного не так. Нам вообще тут jpql не нужен. у нас же есть связь book.getComments, давайте её использовать не обращаясь вообще к репозиторию. Выше уровнями, получили книгу и лениво загрузили комменты

Commit:
Убран лишний импорт

Commit:
Новая версия подгрузки ленивой коллекции
